citro3d/src/texenv.rs

```diff
@@ -3,69 +3,191 @@
+                source1.unwrap_or_default() as _,
+                source2.unwrap_or_default() as _,
+            );
+        }
+
+        if mode.contains(Mode::RGB) {
+            self.sources[0] = source0;
+            self.sources[1] = source1.unwrap_or_default();
+            self.sources[2] = source2.unwrap_or_default();
+        }
+
+        if mode.contains(Mode::ALPHA) {
+            self.sources[3] = source0;
+            self.sources[4] = source1.unwrap_or_default();
+            self.sources[5] = source2.unwrap_or_default();
+        }
+
+        self
+    }
+
+    #[doc(alias = "C3D_TexEnvOpRgb")]
+    pub fn op_rgb(self, o1: RGBOp, o2: Option<RGBOp>, o3: Option<RGBOp>) -> TexEnv {
+        unsafe {
+            citro3d_sys::C3D_TexEnvOpRgb(
+                self.as_raw(),
+                o1 as _,
+                o2.unwrap_or_default() as _,
+                o3.unwrap_or_default() as _,
+            );
+        }
+        self
+    }
+
+    #[doc(alias = "C3D_TexEnvOpAlpha")]
+    pub fn op_alpha(self, o1: AlphaOp, o2: Option<AlphaOp>, o3: Option<AlphaOp>) -> TexEnv {
+        unsafe {
+            citro3d_sys::C3D_TexEnvOpAlpha(
+                self.as_raw(),
+                o1 as _,
+                o2.unwrap_or_default() as _,
+                o3.unwrap_or_default() as _,
             );
         }
         self
     }
 
-    /// Configure the texture combination function.
-    ///
-    /// # Parameters
-    ///
-    /// - `mode`: the [`Mode`]\(s) the combination function will apply to.
-    /// - `func`: the [`CombineFunc`] used to combine textures.
     #[doc(alias = "C3D_TexEnvFunc")]
-    pub fn func(&mut self, mode: Mode, func: CombineFunc) -> &mut Self {
+    pub fn func(self, mode: Mode, func: CombineFunc) -> TexEnv {
         unsafe {
-            citro3d_sys::C3D_TexEnvFunc(self.0, mode.bits(), func as _);
+            citro3d_sys::C3D_TexEnvFunc(self.as_raw(), mode.bits(), func as _);
         }
+        self
+    }
 
+    #[doc(alias = "C3D_TexEnvColor")]
+    pub fn color(self, color: u32) -> TexEnv {
+        unsafe {
+            citro3d_sys::C3D_TexEnvColor(self.as_raw(), color);
+        }
         self
     }
+
+    #[doc(alias = "C3D_TexEnvScale")]
+    pub fn scale(self, mode: Mode, scale: Scale) -> TexEnv {
+        unsafe {
+            citro3d_sys::C3D_TexEnvScale(self.as_raw(), mode.bits() as _, scale as _);
+        }
+        self
+    }
+
+    /// Set this as the active texenv in the given stage.
+    #[must_use]
+    #[doc(alias = "C3D_SetTexEnv")]
+    pub(crate) fn set_texenv(&self, stage: usize) -> crate::Result<()> {
+        if stage >= TEXENV_COUNT {
+            return Err(crate::Error::IndexOutOfBounds {
+                idx: stage as i32,
+                len: TEXENV_COUNT as i32,
+            });
+        }
+
+        unsafe {
+            citro3d_sys::C3D_SetTexEnv(stage as i32, self.as_raw());
+        }
+
+        Ok(())
+    }
```

Comment:
Does this need to have `C3D_DirtyTexEnv` after calling `C3D_SetTexEnv`? I'm not sure whether it's necessary or not. (https://github.com/rust3ds/citro3d-rs/issues/56)

Reply:
No, I don't believe this is necessary. In `C3D_TexEnvDirty` is just setting a flag for the texenv, and in `SetTexEnv` it sets the same flag, so it's basically happening already.